sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx

```diff
@@ -292,29 +307,45 @@ cdef class SparseDenseDatasetsPair(DatasetsPair):
 
     @final
     cdef DTYPE_t surrogate_dist(self, ITYPE_t i, ITYPE_t j) nogil:
+        cdef:
+            ITYPE_t x2_start = j * self.n_features
+            ITYPE_t x2_end = x2_start + self.n_features
+
         return self.distance_metric.rdist_csr(
             x1_data=self.X_data,
-            x1_indices=self.X_indices,
+            x1_indices=&self.X_indices[0],
             x2_data=self.Y_data,
-            x2_indices=self.Y_indices,
+            # To use the same indices array, we shift the array address to map
+            # accesses in [x2_start, x2_end) to
+            # accesses in [x2_start - x2_start, x2_end - x2_start), that is
+            # accesses in [0, n_features).
+            x2_indices=&self.Y_indices[0] - x2_start,
```

Comment:
Two questions:

1. Is this passing a memory address that is invalid?
2. Was this the reason for using pointers instead of memoryviews for the indices?

Reply:
1. Yes, dereferencing this address likely will cause a segmentation fault.
2. Yes, to me using pointers is the only way to use this offset. The implementation of `csr_{rdist,dist}` need not be aware of this. I updated the docstring of one of them to add a remark but since you are asking it might not be totally explicit yet.

Reply:
Even if we do not dereference it, the code does increment the pointer to go back to the "valid memory". Overall, I do not know if incrementing a pointer in this way is defined behavior and I suspect there will be issues.

I opened https://github.com/jjerphan/scikit-learn/pull/16 as an alternative. It increments the data pointer instead of decrementing the indices pointer.